### PR TITLE
mes-3729-removingHardCodedConfigForRateLimiting

### PR DESCRIPTION
**Pull request by amobarak**
Description:

Removed hardcoded values and replaced with environment variables, tested manually in lambda and working fine.

### CODE REVIEW

File: src/functions/sendCandidateResults/framework/request-scheduler.ts [hunk 1]
```diff
@@ -33,13 +33,20 @@ export class RequestScheduler implements IRequestScheduler {
...
-      strategy: bottleneck.strategy.BLOCK, // Ignore any additions to the queue when we reach the max batch size
-      reservoir: 25,                       // Amount of jobs the queue can perform at the start of the queue
-      reservoirRefreshInterval: 1000,      // How often to add new jobs to the queue (every second)
-      reservoirRefreshAmount: 25,          // How many jobs to add to the queue each refresh
+      // No limit on concurrent requests
+      maxConcurrent: null,
+      // No time waited between each request
+      minTime: 0,
+      // Maximum of 250 requests in the queue (max batch size)
+      highWater: this.getNumberFromEnv('NOTIFY_BATCH_SIZE') || 250,
```
Comment (TomBillingtonUK) [inline src/functions/sendCandidateResults/framework/request-scheduler.ts]: Should we not be using the config adapter to set these rather then getting them from the env directly

File: src/functions/sendCandidateResults/framework/request-scheduler.ts [hunk 2]
```diff
@@ -107,6 +114,11 @@ export class RequestScheduler implements IRequestScheduler {
     });
   }
 
+  protected getNumberFromEnv(envvarName: string): number | null {
```
Comment (TomBillingtonUK) [inline src/functions/sendCandidateResults/framework/request-scheduler.ts]: This exists in config adapter

File: src/functions/sendCandidateResults/framework/request-scheduler.ts [hunk 3]
```diff
@@ -33,13 +33,20 @@ export class RequestScheduler implements IRequestScheduler {
...
+      // Maximum of 250 requests in the queue (max batch size)
+      highWater: this.getNumberFromEnv('NOTIFY_BATCH_SIZE') || 250,
+      // Ignore any additions to the queue when we reach the max batch size
+      strategy: bottleneck.strategy.BLOCK,
+      // Amount of jobs the queue can perform at the start of the queue
+      reservoir: this.getNumberFromEnv('NOTIFY_REQUESTS_PER_BATCH') || 25,
+      // How often to add new jobs to the queue (every second)
+      reservoirRefreshInterval: 1000,
+      // How many jobs to add to the queue each refresh
+      reservoirRefreshAmount: this.getNumberFromEnv('NOTIFY_REQUESTS_PER_BATCH') || 25,
```
Comment (rossfauldsbjss) [inline src/functions/sendCandidateResults/framework/request-scheduler.ts]: Should the reservoir and reservoirRefreshAmount always be the same?  
Is it possible we would want these to be different values... e.g the reservoir is 25 but we only want to refresh 15?
If so then they should be two separate environment variables.

**Review by rossfauldsbjss — COMMENTED** [review]
Just need question answered around reservoir and reservoir refresh.

File: src/functions/sendCandidateResults/framework/request-scheduler.ts (hunk 1 above)
Comment (amobarak) [reply src/functions/sendCandidateResults/framework/request-scheduler.ts]: Good point !!!

File: src/functions/sendCandidateResults/framework/request-scheduler.ts (hunk 2 above)
Comment (amobarak) [reply src/functions/sendCandidateResults/framework/request-scheduler.ts]: Yup, will move things from config adapter

File: src/functions/sendCandidateResults/framework/request-scheduler.ts (hunk 3 above)
Comment (ammarhaiderbjss) [reply src/functions/sendCandidateResults/framework/request-scheduler.ts]: Perhaps someone with more knowledge on these fields can weigh in but looking at the tars upload lambda, they are set from the same env var there too - 
https://github.com/dvsa/mes-result-processing-service/blob/develop/src/functions/tarsUpload/framework/adapter/upload/RateLimitDecoratingTARSUploader.ts

Comment (TomBillingtonUK) [reply src/functions/sendCandidateResults/framework/request-scheduler.ts]: From memory, the reservoir is the number of concurrent requests you want to run on initialization and the reservoirRefreshAmount is the number of requests you want to run every interval following the initialization.

In our case, we only want to run 25/second and there is no difference for us between the first and second runs. From the docs =

> reservoir is a counter decremented every time a job is launched, we set its initial value to 100. Then, every reservoirRefreshInterval (60000 ms), reservoir is automatically updated to be equal to the reservoirRefreshAmount (100).

There are some examples of this under Reservoir Intervals in the docs - https://www.npmjs.com/package/bottleneck

Comment (szabikr) [reply src/functions/sendCandidateResults/framework/request-scheduler.ts]: The docs says:
`reservoirRefreshAmount` is the value to set `reservoir` to when `reservoirRefreshInterval` is in use.

**Review by rossfauldsbjss — APPROVED** [review]
looks good

**Review by ammarhaiderbjss — APPROVED** [review]

**Pull request merged**